python/lsst/daf/butler/formatters/fileFormatter.py

```diff
@@ -99,46 +99,54 @@ def _coerceType(self, inMemoryDataset, storageClass, pytype=None):
         """
         return inMemoryDataset
 
-    def read(self, fileDescriptor):
+    def read(self, fileDescriptor, comp=None):
         """Read data from a file.
 
         Parameters
         ----------
         fileDescriptor : `FileDescriptor`
             Identifies the file to read, type to read it into and parameters
             to be used for reading.
+        comp : `str`, optional
+            Component to read from the file. Only used if the `StorageClass`
+            for reading differed from the `StorageClass` used to write the
+            file.
 
         Returns
         -------
         inMemoryDataset : `object`
             The requested data as a Python object. The type of object
             is controlled by the specific formatter.
-        """
-        # Try the file or the component version
-        path = fileDescriptor.location.preferredPath()
-        data = self._readFile(path, fileDescriptor.storageClass.pytype)
-        name = fileDescriptor.location.fragment
-
-        if name:
-            if data is None:
-                # Must be composite written as single file
-                data = self._readFile(fileDescriptor.location.path, fileDescriptor.storageClass.pytype)
 
-                # Now need to "get" the component somehow
-                try:
-                    data = fileDescriptor.storageClass.assembler().getComponent(data, name)
-                except AttributeError:
-                    # Defer the complaint
-                    data = None
+        Raises
+        ------
+        ValueError
+            Component requested but this file does not seem to be a concrete
+            composite.
+        """
 
-            else:
-                # The component was written standalone
-                pass
-        else:
-            # Not requesting a component, so already read
-            pass
+        # Read the file naively
+        path = fileDescriptor.location.path
+        data = self._readFile(path, fileDescriptor.storageClass.pytype)
 
-        data = self._coerceType(data, fileDescriptor.storageClass, pytype=fileDescriptor.pytype)
+        # if read and write storage classes differ, more work is required
+        readStorageClass = fileDescriptor.readStorageClass
+        if readStorageClass != fileDescriptor.storageClass:
+            if comp is None:
+                raise ValueError("Storage class inconsistency ({} vs {}) but no"
+                                 " component requested".format(readStorageClass.name,
+                                                               fileDescriptor.storageClass.name))
+
+            # Concrete composite written as a single file (we hope)
+            try:
+                data = fileDescriptor.storageClass.assembler().getComponent(data, comp)
+            except AttributeError:
```

Comment:
I'm surprised the expected exception is an `AttributeError`; is that a detail of the default implementation?

Reply:
https://github.com/lsst/daf_butler/blob/master/python/lsst/daf/butler/core/composites.py#L205

You prefer `KeyError`? I throw `AttributeError` because I'm testing with `hasattr` in there.